aiidalab_widgets_base/viewers.py

```diff
@@ -766,68 +1083,111 @@ class StructureDataViewer(_StructureDataBaseViewer):
     pk = tl.Int(allow_none=True)
 
     def __init__(self, structure=None, **kwargs):
         super().__init__(**kwargs)
         self.structure = structure
-        self.natom = len(self.structure) if self.structure is not None else 0
 
     @tl.observe("supercell")
-    def repeat(self, _=None):
+    def _observe_supercell(self, _=None):
         if self.structure is not None:
+            self.set_trait(
+                "displayed_structure", None
+            )  # To make sure the structure is always updated.
             self.set_trait("displayed_structure", self.structure.repeat(self.supercell))
 
     @tl.validate("structure")
-    def _valid_structure(self, change):  # pylint: disable=no-self-use
+    def _valid_structure(self, change):
         """Update structure."""
         structure = change["value"]
-
-        if structure is None:
-            return None  # if no structure provided, the rest of the code can be skipped
-
         if isinstance(structure, ase.Atoms):
             self.pk = None
-            return structure
-        if isinstance(structure, orm.Node):
+        elif isinstance(structure, (orm.StructureData, orm.CifData)):
             self.pk = structure.pk
-            return structure.get_ase()
-        raise TypeError(
-            f"Unsupported type {type(structure)}, structure must be one of the following types: "
-            "ASE Atoms object, AiiDA CifData or StructureData."
-        )
+            structure = structure.get_ase()
+
+        # Add default representation array if it is not present.
+        # This will make sure that the new structure is displayed at the beginning.
+        if self.DEFAULT_REPRESENTATION not in structure.arrays:
+            structure.set_array(
+                self.DEFAULT_REPRESENTATION,
+                np.zeros(len(structure), dtype=int),
+            )
+        return structure  # This also includes the case when the structure is None.
 
     @tl.observe("structure")
-    def _observe_structure(self, change):
+    def _observe_structure(self, change=None):
         """Update displayed_structure trait after the structure trait has been modified."""
-        self.natom = len(self.structure) if self.structure is not None else 0
-        # Remove the current structure(s) from the viewer.
-        if change["new"] is not None:
-            self.set_trait("displayed_structure", change["new"].repeat(self.supercell))
-            self.set_trait("cell", change["new"].cell)
-        else:
+        structure = change["new"]
+
+        self._viewer.clear_representations(component=0)
+
+        if not structure:
             self.set_trait("displayed_structure", None)
             self.set_trait("cell", None)
+            return
+
+        # Make sure that the representation arrays from structure are present in the viewer.
+        structure_uuids = [
+            uuid
+            for uuid in structure.arrays
+            if uuid.startswith("_aiidalab_viewer_representation_")
```

Comment:
Should this reference the new constants defined at the top?

Reply:
ah, right, not sure how I missed it. Thanks 👍 
done in https://github.com/aiidalab/aiidalab-widgets-base/pull/373/commits/66ccfbdf2e689b91e927ec2c8b25af3d602f9262